central/auth/datastore/datastore_impl.go

```diff
@@ -72,30 +86,35 @@ func (d *datastoreImpl) UpdateAuthM2MConfig(ctx context.Context, config *storage
 		return err
 	}
 
-	exchanger, err := d.set.NewTokenExchangerFromConfig(ctx, config)
+	ctx, tx, err := d.store.Begin(ctx)
 	if err != nil {
 		return err
 	}
 
-	if err := d.store.Upsert(ctx, config); err != nil {
-		return err
+	exchanger, err := d.set.NewTokenExchangerFromConfig(ctx, config)
+	if err != nil {
+		return d.wrapRollback(ctx, tx, err, existingConfig)
 	}
 
-	if err := d.set.UpsertTokenExchanger(exchanger, config.GetIssuer()); err != nil {
-		return err
+	if err := d.store.Upsert(ctx, config); err != nil {
+		return d.wrapRollback(ctx, tx, err, existingConfig)
 	}
 
+	d.set.UpsertTokenExchanger(exchanger, config.GetIssuer())
+
 	// We need to ensure that any previously existing config is removed from the token exchanger set.
 	// Since this updated config may have updated the issuer, we need to fetch the existing, stored config from the
 	// database and ensure it's removed properly from the set. We do this at the end since we want the new config
 	// to successfully exist beforehand.
 	if exists && config.GetIssuer() != existingConfig.GetIssuer() {
 		if err := d.set.RemoveTokenExchanger(existingConfig.GetIssuer()); err != nil {
-			return err
+			// The removal only returns an error in case the source cannot be unregistered, hence we do not need to
```

Comment:
That might be true for now, but if we change `RemoveTokenExchanger` logic in the future, it can be hard to figure out if this piece of code needs to be modified. Perhaps we can pass `existingConfig.GetIssuer()` anyway?

Reply:
Not really no, because in the rollback parts we are attempting to register the source again. So, in the case of the removal returning an error (i.e. it wasn't possible to unregister the source), in the rollback we attempt to register it once again, which will then return an error because the source already exists.

It is a bit unfortunate, but for the time being I'd live with it. I've added a comment section to the Remove function in the exchanger set to ensure in case things might change over there, it does have implications.